cirq/experiments/fidelity_estimation.py

```diff
@@ -470,17 +505,160 @@ def __enter__(
+            which `GridInteractionLayer` or `Moment` was used for these pairs when calibrating
+            several spacial layouts in one request. This field does not modify any behavior.
+            It is propagated to the output result object.
+        combination_i: Metadata indicating how the `wide_circuit` was zipped. This is
+            the row index of the combinations matrix that identifies this
+            particular combination of component narrow circuits. This field does not modify
+            any behavior. It is propagated to the output result object.
+    """
+
+    wide_circuit: 'cirq.Circuit'
+    pairs: List[Tuple['cirq.Qid', 'cirq.Qid']]
+    combination: List[int]
+    layer_i: int
+    combination_i: int
+
+
+def _get_combinations_by_layer_for_isolated_xeb(
+    circuits: Sequence['cirq.Circuit'],
+) -> Tuple[List[CircuitLibraryCombination], List['cirq.Circuit']]:
+    """Helper function used in `sample_2q_xeb_circuits`.
+
+    This creates a CircuitLibraryCombination object for isolated XEB. First, the qubits
+    are extracted from the lists of circuits and used to define one pair. Instead of using
+    `combinations` to shuffle the circuits for each pair, we just use each circuit (in order)
+    for the one pair.
+    """
+    q0, q1 = _verify_and_get_two_qubits_from_circuits(circuits)
+    circuits = [
+        circuit.transform_qubits(lambda q: {q0: devices.LineQubit(0), q1: devices.LineQubit(1)}[q])
+        for circuit in circuits
+    ]
+    return [
+        CircuitLibraryCombination(
+            layer=None,
+            combinations=np.arange(len(circuits))[:, np.newaxis],
+            pairs=[(q0, q1)],
+        )
+    ], circuits
+
+
+def _zip_circuits(
+    circuits: Sequence['cirq.Circuit'], combinations_by_layer: List[CircuitLibraryCombination]
+) -> List[_ZippedCircuit]:
+    """Helper function used in `sample_2q_xeb_circuits` to zip together circuits.
+
+    This takes a sequence of narrow `circuits` and "zips" them together according to the
+    combinations in `combinations_by_layer`.
+    """
+
+    # Check `combinations_by_layer` is compatible with `circuits`.
+    for layer_combinations in combinations_by_layer:
+        if np.any(layer_combinations.combinations < 0) or np.any(
+            layer_combinations.combinations >= len(circuits)
+        ):
+            raise ValueError("`combinations_by_layer` has invalid indices.")
+
+    zipped_circuits: List[_ZippedCircuit] = []
+    for layer_i, layer_combinations in enumerate(combinations_by_layer):
+        for combination_i, combination in enumerate(layer_combinations.combinations):
+            wide_circuit = Circuit.zip(
+                *(
+                    circuits[i].transform_qubits(lambda q: pair[q.x])
+                    for i, pair in zip(combination, layer_combinations.pairs)
+                )
+            )
+            zipped_circuits.append(
```

Comment:
This and _generate_sample_2q_xeb_tasks could potentialy be generators with `yield`

Reply:
`zipped_circuits` (in `sample_2q_xeb_circuits`) needs to be a list as it's re-iterated-through for each cycle depth. `tasks` needs to be a list as we index into it in order to nest into batches. Therefore, if we change these helper functions to generators an extra `list(...)` would be needed at the call site, so I think it's a wash.